hypertrace-graphql-labels-schema-impl/src/main/java/org/hypertrace/graphql/label/joiner/DefaultLabelJoinerBuilder.java

```diff
@@ -1,58 +1,181 @@
+    this.selectionFinder = selectionFinder;
   }
 
   @Override
   public Single<LabelJoiner> build(GraphQlRequestContext context) {
-    return Single.just(new DefaultLabelJoiner(context));
+    return buildContextualRequest(context)
+        .map(request -> new DefaultLabelJoiner(context, request, null));
+  }
+
+  @Override
+  public Single<LabelJoiner> build(
+      GraphQlRequestContext context, DataFetchingFieldSelectionSet selectionSet) {
+    return zip(
+        buildContextualRequest(context),
+        buildLabeledEntitiesRequestIfPresentInResultSet(selectionSet),
+        (contextualRequest, labeledEntitiesRequests) ->
+            new DefaultLabelJoiner(context, contextualRequest, labeledEntitiesRequests));
+  }
+
+  private Single<ContextualRequest> buildContextualRequest(GraphQlRequestContext context) {
+    return Single.just(this.requestBuilder.build(context));
+  }
+
+  private Single<List<LabeledEntitiesRequest>> buildLabeledEntitiesRequestIfPresentInResultSet(
+      DataFetchingFieldSelectionSet selectionSet) {
+    return Single.just(
+        this.selectionFinder
+            .findSelections(selectionSet, RESULT_SET_LABELED_ENTITIES_SELECTION_QUERY)
+            .map(
+                selectedField ->
+                    new DefaultLabeledEntitiesRequest(
+                        (String) selectedField.getArguments().get(ENTITY_TYPE_ARGUMENT_NAME),
+                        (Integer)
+                            selectedField
+                                .getArguments()
+                                .getOrDefault(LimitArgument.ARGUMENT_NAME, DEFAULT_LIMIT),
+                        selectedField.getSelectionSet()))
+            .collect(Collectors.toList()));
   }
 
   @AllArgsConstructor
   private class DefaultLabelJoiner implements LabelJoiner {
     private final GraphQlRequestContext context;
+    private final ContextualRequest resultSetRequest;
+    private final List<LabeledEntitiesRequest> labeledEntitiesRequests;
 
     @Override
     public <T> Single<Map<T, LabelResultSet>> joinLabels(
         Collection<T> joinSources, LabelIdGetter<T> labelIdGetter) {
       return labelDao
-          .getLabels(requestBuilder.build(context))
+          .getLabels(resultSetRequest)
           .flatMap(labelResultSet -> buildLabelMap(labelResultSet.results()))
-          .flatMap(labelMap -> getLabelResultSetMap(joinSources, labelIdGetter, labelMap));
+          .flatMap(labelsMap -> getLabelResultSetMap(joinSources, labelIdGetter, labelsMap));
+    }
+
+    @Override
+    public Single<LabelResultSet> joinLabelsWithEntities() {
+      return labelDao.getLabels(resultSetRequest).flatMap(this::getUpdatedLabelResultSet);
+    }
+
+    private Single<LabelResultSet> getUpdatedLabelResultSet(LabelResultSet labelResultSet) {
+      return getUpdatedLabels(labelResultSet).flatMap(responseConverter::convert);
+    }
+
+    private Single<List<Label>> getUpdatedLabels(LabelResultSet labelResultSet) {
```

Comment:
To my earlier point, using the joiner in place of the DAO is what leads to this strange fetch, build response then update pattern. Ideally, we're doing the entity joining before constructing the label response.

Reply:
Since dao returns a result set, you need to something similar isn't it? Even in entity joiner, entity dao returns entity result set which is then converted to a table and then back to appropriate result set.

Reply:
in the current model, some other dao (say, label), would use the joiner when constructing its objects (labels). The entity joiner calls the entity dao and presents the results. Right now, that means its unpacking the result set into a table, and in the version we have here you may need to unpack and repack the result sets. Ideally that's not necessary and we can use them as is, but I think that may be required to enforce limits that differ.